### ISSUE TITLE
R code

### ISSUE THREAD

**Issue by 3laasameh** [3laasameh/alaa#3]
`# ***Installing packages ***
install.packages(c("tidyverse", "ggplot2", "cluster", "factoextra",
                   "caret", "randomForest", "DataExplorer", "iml"))
library(tidyverse)
library(ggplot2)
library(cluster)
library(factoextra)
library(caret)
library(randomForest)
library(DataExplorer)
library(iml)
loading data at first
view the first few rows as default
then check the structure of our data
# Load the data
data <- read.csv("breast cancer data.csv", header = FALSE)  # header = FALSE in case the file doesn't have column names

# View the first few rows
head(data)

# Check the structure
str(data)

colnames(data) <- c("ID", "Diagnosis", "Radius_mean", "Texture_mean", "Perimeter_mean",
                    "Area_mean", "Smoothness_mean", "Compactness_mean", "Concavity_mean",
                    "Concave_points_mean", "Symmetry_mean", "Fractal_dimension_mean",
                    "Radius_se", "Texture_se", "Perimeter_se", "Area_se", "Smoothness_se",
                    "Compactness_se", "Concavity_se", "Concave_points_se", "Symmetry_se",
                    "Fractal_dimension_se", "Radius_worst", "Texture_worst", "Perimeter_worst",
                    "Area_worst", "Smoothness_worst", "Compactness_worst", "Concavity_worst",
                    "Concave_points_worst", "Symmetry_worst", "Fractal_dimension_worst")

# 1-reading the data
# 2-convert to appropiate types
library(tibble)
library(dplyr)

# Step 1: Read data
raw_data <- read.csv("breast cancer data.csv", header = FALSE, stringsAsFactors = FALSE)
col_names <- as.character(unlist(raw_data[1, ]))
raw_data <- raw_data[-1, ]
data <- as_tibble(matrix(unlist(raw_data), ncol = length(col_names), byrow = FALSE))
colnames(data) <- col_names

# Step 2: Convert to appropriate types
data <- data %>%
  mutate(across(where(is.character), ~ trimws(.))) %>%  # Trim whitespace
  mutate(across(c(-id, -diagnosis), as.numeric)) %>%    # Convert all but 'id' and 'diagnosis' to numeric
  mutate(
    id = as.factor(id),
    diagnosis = as.factor(tolower(diagnosis))           # Lowercase 'diagnosis' values
  ) %>%
  filter(if_all(where(is.numeric), ~ !is.na(.) & . >= 0))  # Remove rows with negative or NA numeric values

# **we will go toe remove the outliers and compute the missing values using median **
library(dplyr)

# Remove outliers above the 99th percentile in any numeric column
data <- data %>%
  filter(if_all(where(is.numeric), ~ . < quantile(., 0.99, na.rm = TRUE)))

# Impute missing values in numeric columns using median
data <- data %>%
  mutate(across(where(is.numeric), ~ ifelse(is.na(.), median(., na.rm = TRUE), .)))
# ploting the data structure
# Load required libraries
library(ggplot2)
library(dplyr)
library(corrplot)
library(grid)

# Function to plot with space and explanation
plot_with_explanation <- function(plot_obj, explanation_text) {
  print(plot_obj)
  grid.text(explanation_text,
            x = 0.5, y = unit(1, "npc") - unit(5, "lines"), just = "center",
            gp = gpar(fontsize = 11, col = "darkblue"))
  grid.newpage()  # add space before next plot
}

# Assuming 'data' is your dataframe with breast cancer data loaded

# 1. Diagnosis distribution plot
p1 <- ggplot(data, aes(x = diagnosis, fill = diagnosis)) +
  geom_bar() +
  theme_minimal() +
  labs(title = "Diagnosis Distribution", x = "Diagnosis", y = "Count") +
  scale_fill_manual(values = c("benign" = "skyblue", "malignant" = "salmon"))

# 2. Histogram of radius_mean by diagnosis
p2 <- ggplot(data, aes(x = radius_mean, fill = diagnosis)) +
  geom_histogram(bins = 30, alpha = 0.6, position = "identity", color = "black") +
  theme_minimal() +
  labs(title = "Distribution of Radius Mean by Diagnosis", x = "Radius Mean", y = "Count")

# 3. Scatter plot radius_mean vs area_mean colored by diagnosis
p3 <- ggplot(data, aes(x = radius_mean, y = area_mean, color = diagnosis)) +
  geom_point(alpha = 0.6) +
  theme_minimal() +
  labs(title = "Radius Mean vs Area Mean", x = "Radius Mean", y = "Area Mean")

# 4. Correlation matrix plot of numeric features
numeric_data <- select_if(data, is.numeric)
cor_matrix <- cor(numeric_data, use = "complete.obs")

# Now print each plot with its explanation and space
plot_with_explanation(p1, explanation1)
plot_with_explanation(p2, explanation2)
plot_with_explanation(p3, explanation3)

# Plot correlation matrix
corrplot(cor_matrix, method = "color", type = "upper", tl.cex = 0.7, title = "Correlation Matrix", mar = c(0,0,1,0))

# Print explanation for correlation matrix in console
cat("Correlation matrix shows strength and direction of linear relationships between numeric features.\n")

box plot to check if there is outliers

# Boxplot of radius_mean by diagnosis
p4 <- ggplot(data, aes(x = diagnosis, y = radius_mean, fill = diagnosis)) +
  geom_boxplot() +
  theme_minimal() +
  labs(title = "Boxplot of Radius Mean by Diagnosis", x = "Diagnosis", y = "Radius Mean") +
  scale_fill_manual(values = c("benign" = "skyblue", "malignant" = "salmon"))

explanation4 <- "Boxplot shows the distribution (median, quartiles, and outliers) of 'radius_mean' values for each diagnosis group."

# Now plot it with explanation and spacing
plot_with_explanation(p4, explanation4)

# Previous plots here...
plot_with_explanation(p1, explanation1)
plot_with_explanation(p2, explanation2)
plot_with_explanation(p3, explanation3)

# Boxplot
plot_with_explanation(p4, explanation4)

# Then correlation matrix plot
corrplot(cor_matrix, method = "color", type = "upper", tl.cex = 0.7, title = "Correlation Matrix", mar = c(0,0,1,0))
cat("Correlation matrix shows strength and direction of linear relationships between numeric features.\n")

library(dplyr)
library(factoextra)  # For visualization

set.seed(123)

# Sample up to 1000 rows (or all if less than 1000)
sampled_data <- data %>% sample_n(min(1000, nrow(data)))

# Select numeric features for clustering
cluster_data <- sampled_data %>%
  select(radius_mean, texture_mean, perimeter_mean, area_mean, smoothness_mean) %>%
  scale()

# Determine the optimal number of clusters using Elbow method
fviz_nbclust(cluster_data, kmeans, method = "wss") +
  ggtitle("Elbow Method for Choosing k")

# Apply k-means with k = 3 (or choose k based on the elbow plot)
kmeans_model <- kmeans(cluster_data, centers = 3, nstart = 25)

# Add cluster assignments to the sampled_data
sampled_data$Cluster <- as.factor(kmeans_model$cluster)

# Define your custom colors for clusters, e.g., 3 clusters:
my_colors <- c("#E41A1C", "#377EB8", "#4DAF4A")  # red, blue, green
# Visualize clusters with custom colors
fviz_cluster(kmeans_model, data = cluster_data, geom = "point", ellipse.type = "norm") +
  ggtitle("K-means Clustering with Custom Colors") +
  scale_color_manual(values = my_colors) +
  scale_fill_manual(values = my_colors)
set.seed(123)

# Example dataset with numeric features and a factor target
model_data <- data.frame(
  Age = sample(20:70, 200, replace = TRUE),
  Stay_Length = sample(1:30, 200, replace = TRUE),
  Billing.Amount = runif(200, 100, 5000),
  Test.Results = sample(c("Positive", "Negative"), 200, replace = TRUE)
)
library(caret)

set.seed(123)
trainIndex <- createDataPartition(model_data$Test.Results, p = 0.8, list = FALSE)
train <- model_data[trainIndex, ]
test <- model_data[-trainIndex, ]

# Ensure reticulate is installed and loaded if you plan to run Python from R
# install.packages("reticulate")
# library(reticulate)

# Use reticulate's function to install Python packages
reticulate::py_install(packages = c("tensorflow", "keras", "numpy"), pip = TRUE)
# Train-test split
set.seed(123)
trainIndex <- createDataPartition(model_data$Test.Results, p = 0.8, list = FALSE)
train <- model_data[trainIndex, ]
test <- model_data[-trainIndex, ]
# Ensure the randomForest package is loaded
# install.packages("randomForest") # Uncomment and run if you suspect the package is not installed
library(randomForest)
library(caret) # Ensure caret is loaded for createDataPartition

# Example dataset with numeric features and a factor target
# (Keeping this as per the original notebook, but note that small datasets increase the chance of a single-level split)
model_data <- data.frame(
  Age = sample(20:70, 200, replace = TRUE),
  Stay_Length = sample(1:30, 200, replace = TRUE),
  Billing.Amount = runif(200, 100, 5000),
  Test.Results = sample(c("Positive", "Negative"), 200, replace = TRUE)
)

# Train-test split
set.seed(123) # You can remove or change the seed to get a different split
trainIndex <- createDataPartition(model_data$Test.Results, p = 0.8, list = FALSE)
train <- model_data[trainIndex, ]
test <- model_data[-trainIndex, ]

# Explicitly ensure the target variable is a factor (should already be, but re-confirming)
train$Test.Results <- as.factor(train$Test.Results)

# Check the number of levels in the target variable in the training data
print(levels(train$Test.Results))
print(paste("Number of levels in train$Test.Results:", nlevels(train$Test.Results)))

# Train Random Forest model
# This step will succeed if train$Test.Results has more than one level
rf_model <- randomForest(Test.Results ~ ., data = train, importance = TRUE)

print("Random Forest model trained successfully.")
# Ensure the randomForest package is loaded
# install.packages("randomForest") # Uncomment and run if you suspect the package is not installed
library(randomForest)
library(caret) # Ensure caret is loaded for createDataPartition

# Example dataset with numeric features and a factor target
# (Keeping this as per the original notebook, but note that small datasets increase the chance of a single-level split)
model_data <- data.frame(
  Age = sample(20:70, 200, replace = TRUE),
  Stay_Length = sample(1:30, 200, replace = TRUE),
  Billing.Amount = runif(200, 100, 5000),
  Test.Results = sample(c("Positive", "Negative"), 200, replace = TRUE)
)

# Train-test split
set.seed(123) # You can remove or change the seed to get a different split
trainIndex <- createDataPartition(model_data$Test.Results, p = 0.8, list = FALSE)
train <- model_data[trainIndex, ]
test <- model_data[-trainIndex, ]

# Explicitly ensure the target variable is a factor (should already be, but re-confirming)
train$Test.Results <- as.factor(train$Test.Results)

# Check the number of levels in the target variable in the training data
print(levels(train$Test.Results))
print(paste("Number of levels in train$Test.Results:", nlevels(train$Test.Results)))

# Train Random Forest model
# This step will succeed if train$Test.Results has more than one level
rf_model <- randomForest(Test.Results ~ ., data = train, importance = TRUE)

print("Random Forest model trained successfully.")
predictions <- predict(rf_model, newdata = test)

# Convert predictions to character, then back to a factor with the desired levels
# This handles cases where the predictions might not include all levels present in the test data
predictions <- factor(as.character(predictions), levels = levels(test$Test.Results))

# Now you can proceed with evaluation, e.g.,
# confusionMatrix(predictions, test$Test.Results)
# Ensure the randomForest package is loaded
# install.packages("randomForest") # Uncomment and run if you suspect the package is not installed
library(randomForest)
library(caret) # Ensure caret is loaded for createDataPartition

# Example dataset with numeric features and a factor target
# (Keeping this as per the original notebook, but note that small datasets increase the chance of a single-level split)
model_data <- data.frame(
  Age = sample(20:70, 200, replace = TRUE),
  Stay_Length = sample(1:30, 200, replace = TRUE),
  Billing.Amount = runif(200, 100, 5000),
  Test.Results = sample(c("Positive", "Negative"), 200, replace = TRUE)
)

# Train-test split
set.seed(123) # You can remove or change the seed to get a different split
trainIndex <- createDataPartition(model_data$Test.Results, p = 0.8, list = FALSE)
train <- model_data[trainIndex, ]
test <- model_data[-trainIndex, ]

# Explicitly ensure the target variable is a factor (should already be, but re-confirming)
train$Test.Results <- as.factor(train$Test.Results)
test$Test.Results <- as.factor(test$Test.Results) # Ensure test target is also a factor

# Check the number of levels in the target variable in the training data
print(levels(train$Test.Results))
print(paste("Number of levels in train$Test.Results:", nlevels(train$Test.Results)))

# Train Random Forest model
# This step will succeed if train$Test.Results has more than one level
rf_model <- randomForest(Test.Results ~ ., data = train, importance = TRUE)

print("Random Forest model trained successfully.")

# Get predictions
predictions <- predict(rf_model, newdata = test)

# Get all unique levels from both predictions and the actual test results
all_levels <- unique(c(levels(predictions), levels(test$Test.Results)))

# Coerce both predictions and test$Test.Results to factors with the unified levels
predictions_factor <- factor(as.character(predictions), levels = all_levels)
test_results_factor <- factor(as.character(test$Test.Results), levels = all_levels)

# Now calculate and show the confusion matrix and accuracy using caret

# Ensure the caret package is loaded
# install.packages("caret") # Uncomment if not installed
library(caret)

# Calculate the confusion matrix using the factors with aligned levels
conf_matrix <- confusionMatrix(predictions_factor, test_results_factor)

# Print the confusion matrix
print(conf_matrix)

# Extract and print the accuracy
accuracy <- conf_matrix$overall['Accuracy']
print(paste("Accuracy:", accuracy))
# Ensure the randomForest package is loaded
# install.packages("randomForest") # Uncomment and run if you suspect the package is not installed
# library(randomForest) # We won't need randomForest for this SVM example, but keeping for context

# Ensure e1071 and caret packages are loaded
# install.packages("e1071") # Uncomment if not installed
# install.packages("caret") # Uncomment if not installed
library(e1071)
library(caret)

# Example dataset with numeric features and a factor target
# (Keeping this as per the original notebook)
model_data <- data.frame(
  Age = sample(20:70, 200, replace = TRUE),
  Stay_Length = sample(1:30, 200, replace = TRUE),
  Billing.Amount = runif(200, 100, 5000),
  Test.Results = sample(c("Positive", "Negative"), 200, replace = TRUE)
)

# Train-test split
set.seed(123) # You can remove or change the seed to get a different split
trainIndex <- createDataPartition(model_data$Test.Results, p = 0.8, list = FALSE)
train <- model_data[trainIndex, ]
test <- model_data[-trainIndex, ]

# Explicitly ensure the target variable is a factor
train$Test.Results <- as.factor(train$Test.Results)
test$Test.Results <- as.factor(test$Test.Results) # Ensure test target is also a factor

# --- Train SVM model ---
# We use the `svm` function from the e1071 package
# kernel = "linear" is a common starting point; other options include "radial", "polynomial", etc.
# cost = 1 is the regularization parameter
svm_model <- svm(Test.Results ~ ., data = train, kernel = "linear", cost = 1)

print("SVM model trained successfully.")

# --- Make predictions ---
# Make predictions on the test data
predictions <- predict(svm_model, newdata = test)

# --- Evaluate the model using a Confusion Matrix ---
# Get all unique levels from both predictions and the actual test results
# This step is crucial to avoid the 'levels' error
all_levels <- unique(c(levels(predictions), levels(test$Test.Results)))

# Coerce both predictions and test$Test.Results to factors with the unified levels
predictions_factor <- factor(as.character(predictions), levels = all_levels)
test_results_factor <- factor(as.character(test$Test.Results), levels = all_levels)

# Calculate the confusion matrix using the factors with aligned levels
conf_matrix <- confusionMatrix(predictions_factor, test_results_factor)

# Print the confusion matrix
print(conf_matrix)

# Extract and print the accuracy
accuracy <- conf_matrix$overall['Accuracy']
print(paste("Accuracy:", accuracy))`
